types/backbone/index.d.ts

```diff
@@ -410,73 +407,62 @@ declare namespace Backbone {
-        isEmpty(): boolean;
-        last(): TModel;
-        last(n: number): TModel[];
-        lastIndexOf(value: TModel, from?: number): number;
-        map<TResult>(iterator: _.ListIterator<TModel, TResult>, context?: any): TResult[];
-        max(iterator?: _.ListIterator<TModel, any>, context?: any): TModel;
-        min(iterator?: _.ListIterator<TModel, any>, context?: any): TModel;
-        partition(iterator: _.ListIterator<TModel, boolean>): TModel[][];
-        reduce<TResult>(iterator: _.MemoIterator<TModel, TResult>, memo?: TResult, context?: any): TResult;
-        reduceRight<TResult>(iterator: _.MemoIterator<TModel, TResult>, memo?: TResult, context?: any): TResult;
-        reject(iterator: _.ListIterator<TModel, boolean>, context?: any): TModel[];
-        rest(n?: number): TModel[];
-        sample(): TModel;
-        sample(n: number): TModel[];
-        select(iterator: _.ListIterator<TModel, boolean>, context?: any): TModel[];
-        shuffle(): TModel[];
-        size(): number;
-        some(iterator?: _.ListIterator<TModel, boolean>, context?: any): boolean;
-        sortBy<TSort>(iterator?: _.ListIterator<TModel, TSort>, context?: any): TModel[];
-        sortBy(iterator: string, context?: any): TModel[];
-        tail(n?: number): TModel[];
-        take(): TModel;
-        take(n: number): TModel[];
-        toArray(): TModel[];
+        all: _.Underscore<TModel>['all'];
+        any: _.Underscore<TModel>['any'];
+        chain: _.Underscore<TModel>['chain'];
+        collect: _.Underscore<TModel>['collect'];
+        contains: _.Underscore<TModel>['contains'];
+        countBy: _.Underscore<TModel>['countBy'];
+        detect: _.Underscore<TModel>['detect'];
+        difference: _.Underscore<TModel>['difference'];
+        drop: _.Underscore<TModel>['drop'];
+        each: _.Underscore<TModel>['each'];
+        every: _.Underscore<TModel>['every']
+        filter: _.Underscore<TModel>['filter'];
+        find: _.Underscore<TModel>['find'];
+        findIndex: _.Underscore<TModel>['findIndex'];
+        findLastIndex: _.Underscore<TModel>['findLastIndex'];
+        first: _.Underscore<TModel>['first'];
+        foldl: _.Underscore<TModel>['foldl'];
+        foldr: _.Underscore<TModel>['foldr'];
+        forEach: _.Underscore<TModel>['forEach'];
+        groupBy: _.Underscore<TModel>['groupBy'];
+        head: _.Underscore<TModel>['head'];
+        include: _.Underscore<TModel>['include'];
+        includes: _.Underscore<TModel>['includes'];
+        indexBy: _.Underscore<TModel>['indexBy'];
+        indexOf: _.Underscore<TModel>['indexOf'];
+        initial: _.Underscore<TModel>['initial'];
+        inject: _.Underscore<TModel>['inject'];
+        invoke: _.Underscore<TModel>['invoke'];
+        isEmpty: _.Underscore<TModel>['isEmpty'];
+        last: _.Underscore<TModel>['last'];
+        lastIndexOf: _.Underscore<TModel>['lastIndexOf'];
+        map: _.Underscore<TModel>['map'];
```

Comment:
`map` and `collect` are a bit interesting because they look at the properties of the current object when passed a string argument, so if I were to add an attribute `foo` to a model and then also add a `foo` property to that model that was of a different type than the attribute, type inference would end up picking the wrong type for the result. It might be worth making changes to `_.Iteratee` to allow the resulting type for a string (and maybe later path) argument to be overridden somehow to deal with that possibility, though on the other hand that probably doesn't happen super often since it seems like it would potentially be confusing.

Reply:
I think I may need to make a change like this for the build failures that are happening as well - a matcher iteratee type of `Partial<TModel>` is also not accurate for these functions.

Reply:
There's something sophisticated going on here; Backbone is [wrapping](https://backbonejs.org/docs/backbone.html#section-186) the mixed methods so that all iteratee shorthands are applied to the model's `attributes` object instead of to the model object as a whole. I think you're aware of that, judging by what you wrote, but it means that the types are different from stock Underscore for all collection and array functions. You can, however, write a metafunction that maps the original type to the mixed Backbone type (I think).